### PR TITLE
*: make auto-analyze killable by global memory limit

### LINKED ISSUE #39994: auto analyze job cannot be killed by tidb_server_memory_limit
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. set global memory limit to 6G
2. consume 5G long-existing memory
3. trigger an auto analyze who consumes 1G and appears as the top 1 memory consumer

### 2. What did you expect to see? (Required)
auto analyze got killed

### 3. What did you see instead (Required)
auto analyze cannot be killed

### 4. What is your TiDB version? (Required)
v6.5

### PR DESCRIPTION
**Pull request by chrysan**
### What problem does this PR solve?

Issue Number: close #39994 & #39971

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: testkit/mocksessionmanager.go [hunk 1]
```diff
@@ -31,6 +32,7 @@ type MockSessionManager struct {
 	PSMu    sync.RWMutex
 	SerID   uint64
 	TxnInfo []*txninfo.TxnInfo
+	DOM     *domain.Domain
```
Comment (xuyifangreeneyes) [inline testkit/mocksessionmanager.go:35]:
```suggestion
	Dom     *domain.Domain
```

File: util/memory/tracker.go [hunk 2]
```diff
@@ -762,6 +762,17 @@ func (t *Tracker) CountAllChildrenMemUse() map[string]int64 {
 	return trackerMemUseMap
 }
 
+// GetChildren returns children trackers
+func (t *Tracker) GetChildren() []*Tracker {
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:766]: GetChildrenForTest

**Review by XuHuaiyu — APPROVED** [review]

File: executor/executor_test.go [hunk 3]
```diff
@@ -6220,6 +6221,60 @@ func TestGlobalMemoryControlForAnalyze(t *testing.T) {
...
+	}()
+	h.HandleAutoAnalyze(dom.InfoSchema())
+	rs := tk0.MustQuery("select fail_reason from mysql.analyze_jobs where table_name=? and state=? limit 1", "t", "failed")
+	failReason := rs.Rows()[0][0].(string)
+	require.True(t, strings.Contains(failReason, "Out Of Memory Quota!"))
+
+	// There should be only one child tracker, auto-analyze session should have been detached.
+	childTrackers := executor.GlobalAnalyzeMemoryTracker.GetChildrenForTest()
+	require.Len(t, childTrackers, 1)
+	require.Equal(t, tk0.Session().GetSessionVars().MemTracker, childTrackers[0])
```
Comment (xuyifangreeneyes) [inline executor/executor_test.go:6275]: If some user session has executed the analyze command, then the session's MemTracker is attached to `GlobalAnalyzeMemoryTracker` and the memory usage of the following commands executed in the session would count in global analyze memory.

Comment (chrysan) [reply executor/executor_test.go]: You'are right, fix it.

**Review by xuyifangreeneyes — APPROVED** [review]

File: testkit/mocksessionmanager.go [hunk 4]
```diff
@@ -85,6 +87,11 @@ func (msm *MockSessionManager) GetProcessInfo(id uint64) (*util.ProcessInfo, boo
 	if sess := msm.conn[id]; sess != nil {
 		return sess.ShowProcess(), true
 	}
+	if msm.Dom != nil {
+		if pinfo, ok := msm.Dom.SysProcTracker().GetSysProcessList()[id]; ok {
+			return pinfo, true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (chrysan) [reply testkit/mocksessionmanager.go:99]: Added

**Comment by hawkingrei** [comment]
```
analyze_test.go:3147: 

        	Error Trace:	/home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/7606/execroot/__main__/bazel-out/k8-fastbuild/bin/executor/analyzetest/analyzetest_test_/analyzetest_test.runfiles/__main__/executor/analyzetest/analyze_test.go:3147

        	Error:      	"[

        	            	"-27"{

        	            	  "consumed": 0 Bytes

        	            	}

        	            	]" should have 0 item(s), but has 1

        	Test:       	TestGlobalMemoryControlForAutoAnalyze

```

**Pull request merged**